lib/_internal/_set/index.js

```diff
@@ -0,0 +1,53 @@
+// TODO: to complete for objects with recursion
+function getHashCode(object) { 
+  return (typeof object === 'function' 
+      ? object + '' 
+      : JSON.stringify(object)
+    )
+    .split('')
+    .reduce(function (hash, value) {
+      var nHash = ((hash << 5) - hash) + value.charCodeAt(0)
+      return nHash & nHash
+    }, 0)
+}
+
+function has(bucket, element) {
+  var stringified = JSON.stringify(element)
+  return (bucket[getHashCode(element)] || []).some(function (e) {
+    return e === stringified
+  })
+}
+
+function add(bucket, element) {
+  var hash = getHashCode(element)
+  bucket[hash] = (bucket[hash] || []).concat(JSON.stringify(element))
+  return bucket
+}
+
+function updateLocals(array, set, values) {
+  array.forEach(function (element) {
+    if (!has(set, element)) {
+      add(set, element)
+      values.push(element)
```

Comment:
It's not mistake, but there is the question: are we totally using only FP or not?

Reply:
yeah, but below I replace values and return new array, see https://github.com/nanoutils/nanoutils/pull/10/files#diff-526f1bc96323877f12015beff60fb247R50

Reply:
So, it's only internally mutable array?

Reply:
@BabkinAleksandr `_set` returns the object with custom methods, it doesn't know about inner variables and they are not accessible from outside. `concat` returns `this` as well, `values` returns `Array` of values, a copy of inner `values`. `has` returns `Boolean`. No mutation at all.

Reply:
`updateLocals` function mutates provided `values` array, that stored in `_set` function. Than `_set` return new copy of that `values` array.

So `_set` stores 'values` array and mutates it for every change. It could be said about `set` object.

I mean it's mutating, but it's not a mistake or big deal, just mention it.

Reply:
@BabkinAleksandr yeah, it should be kept in mind